mono/mini/interp/interp.c

```diff
@@ -5139,17 +5136,10 @@ call:;
 				sp [1].data.o = o;
 				sp += param_count + 2;
 			} else {
-				InterpMethod *ctor_method = (InterpMethod*) frame->imethod->data_items [imethod_index];
+				cmethod = (InterpMethod*)frame->imethod->data_items [imethod_index];
 				frame->ip = ip;
-
-				child_frame = alloc_frame (context, &vtable, frame, ctor_method, sp, NULL);
-
-				// FIXME Remove recursion.
-				interp_exec_method_newobj_fast (child_frame, context, error);
-
-				CHECK_RESUME_STATE (context);
-				sp [0].data.o = o;
-				sp++;
+				retval = NULL;
+				goto call;
```

Comment:
I wonder if you can do something similar to what the inlined call case is doing: store an extra copy of `o` on the data stack before the call, and then in `exit_frame:` you wouldn't need to do anything special - returning from the constructor will pop off `this` and the constructor arguments, and the value left on the top of the stack will be `o`.  

That was you wouldn't have to store `o` in the interpreter state and the whole question of the GC hole can be avoided.

Reply:
I was wondering that too — to use stack.
 Probably need a larger stack by 1.
The grow-up stack hurts my head, I too used to x86.

Reply:
Or is it larger by one for each newobj_fast, in case they nest?
`newobj_fast(newobj_fast())` etc.

Reply:
yes, I think you need to also do some PUSH_TYPE stuff in `transform.c` so that max_stack_height is correct.

There wouldn't be nesting.   `new Foo (new Bar (42))` would become something like: `push 42; newobj_fast Bar; newobj_fast  Foo`.  So then you want it to behave like this pseudo-Forth:

````
push 42             //  42
alloc sizeof(Bar) // 42 o1
dup              // 42 o1 o1
rot               // o1 42 o1
call Bar::.ctor;  // pops 2 args; on return left with just o1
alloc sizeof(Foo) // o1 o2
dup         // o1 o2 o2
rot         // o2 o1 o2
call Foo::.ctor // pops 2 args, on return left with o2
```

(where `rot` behaves like this `a b c ---> c a b`)

where the `dup; rot` sequence is basically a `memmove` and a write to the freed up slot.

Reply:
Actually it's possible that transform.c is already good.  https://github.com/mono/mono/blob/b2ad5f34e18a8b01f7c384221cc224a4d7680bb8/mono/mini/interp/transform.c#L4505-L4509

looks like it reserves 2 extra slots regardless of whether the constructor will be inlined or not.

Reply:
What about a few lines later:

https://github.com/mono/mono/blob/b2ad5f34e18a8b01f7c384221cc224a4d7680bb8/mono/mini/interp/transform.c#L4520

Reply:
That part looks like, at least size-wise.
Negative moves do not shink the stack.
The types might need attention though.
Maybe remove that line? Or change -2 to -1?
Or change `*push = 1;` or `*push = 2;?` in:
```
		case MINT_NEWOBJ_FAST: {
			int param_count = ins->data [1];
			gboolean is_inlined = ins->data [0] == INLINED_METHOD_FLAG;
			if (is_inlined) {
				// This needs to be handled explictly during cprop, in order to properly
				// keep track of stack contents
				*pop = 0;
				*push = 2;
			} else {
				*pop = param_count;
				*push = 1;
			}
			break;
		}
```

Reply:
Well, it is still push1 in the end, just temporarily more, so probably not that last thing.